File: services/iam/internal/app/passwordpolicy.go
package app

import (
	"strings"
	"unicode"

	"github.com/sgao19/erp-go/pkg/apierr"
)

// What makes a password acceptable.
//
// This file exists because of a trade-off made in the rate limiter next door.
// The lockout window was cut from fifteen minutes to sixty seconds so that
// somebody being attacked is shut out for a minute rather than a quarter of an
// hour — but a shorter lock means more guesses per hour get through, and
// ERPNext, whose shape that change copies, only affords a loose limit because
// it ships password scoring switched on by default.
//
// So the two move together. Loosening the limit without this would be a strict
// downgrade, and shipping only one half of the pair is the kind of change that
// looks like hardening and is not.
//
// What is deliberately NOT here: composition rules. Requiring an upper-case
// letter, a digit and a symbol is how you get "P@ssw0rd1" — it converts the
// whole workforce onto the same handful of predictable shapes, which is the
// opposite of the goal. Length, plus refusing the passwords attackers actually
// try, does more.

const (
	// Ten, not eight. Eight is the floor NIST sets for a system that also
	// rate-limits hard; ours now rate-limits gently, so the password carries
	// more of the weight.
	//
	// Measured in effective length rather than characters — see
	// effectiveLength, which exists because a character-count minimum quietly
	// punishes the language most of this company writes in.
	minPasswordLen = 10
	// And a floor on actual characters, whatever alphabet they come from. Five
	// hanzi carry plenty of theoretical entropy and are also how you write a
	// four-word phrase everybody knows.
	minPasswordRunes = 6
	// No upper bound worth enforcing below this. A long passphrase is the
	// best password most people will ever choose, and a maximum length is a
	// reason not to use one. This exists only so a megabyte of text cannot be
	// pushed through argon2.
	maxPasswordLen = 256
)

// commonPasswords is the list attackers work through first.
//
// Weighted towards what is actually tried against a Chinese company rather
// than a generic English top-100: keyboard walks, birthday-shaped digits,
// 拼音 endearments, and the "admin/test/1qaz" family every scanner carries.
// A short curated list catches most real attempts; the tail is long and adds
// little, which is why this is a compiled-in set rather than a data file
// somebody has to remember to ship.
var commonPasswords = map[string]bool{
	"password": true, "passwd": true, "pass": true, "admin": true, "administrator": true,
	"root": true, "test": true, "guest": true, "user": true, "login": true,
	"welcome": true, "letmein": true, "monkey": true, "dragon": true, "master": true,
	"qwerty": true, "qwertyuiop": true, "asdfgh": true, "asdfghjkl": true, "zxcvbn": true,
	"1qaz2wsx": true, "1q2w3e4r": true, "1q2w3e": true, "qazwsx": true, "qweasd": true,
	"123456": true, "1234567": true, "12345678": true, "123456789": true, "1234567890": true,
	"111111": true, "000000": true, "666666": true, "888888": true, "123123": true,
	"abc123": true, "a123456": true, "123abc": true, "abcd1234": true, "qwe123": true,
	"woaini": true, "woaini1314": true, "5201314": true, "1314520": true, "wangyi": true,
	"iloveyou": true, "sunshine": true, "princess": true, "football": true, "baseball": true,
	"changeme": true, "secret": true, "default": true, "temp": true, "demo": true,
}

// checkPasswordStrength refuses what a person will regret choosing.
//
// context is whatever the attacker already knows about this account — the
// address, the person's name, the company domain. Those are the first things
// tried and the first things people reach for, so a password built out of them
// is the worst of both.
func checkPasswordStrength(password string, context ...string) error {
	runes := []rune(password)
	if len(runes) < minPasswordRunes || effectiveLength(runes) < minPasswordLen {
		return apierr.Invalid("IAM_PASSWORD_TOO_SHORT", "密码太短了，至少 10 位（中文可以短一些）")
	}
	if len(runes) > maxPasswordLen {
		return apierr.Invalid("IAM_PASSWORD_TOO_LONG", "密码太长了，请控制在 256 位以内")
	}
	if isCommonPassword(password) || walksTheKeyboard(password) {
		return apierr.Invalid("IAM_PASSWORD_TOO_COMMON",
			"这是最常被猜到的密码之一，换一个")
	}
	if isRepetitive(runes) {
		return apierr.Invalid("IAM_PASSWORD_TOO_SIMPLE",
			"密码太规律了（重复或连续的字符），换一个")
	}
	if who := borrowsFromIdentity(password, context); who != "" {
		return apierr.Invalid("IAM_PASSWORD_FROM_IDENTITY",
			"密码里不要包含「"+who+"」——猜密码的人第一个就试这个")
	}
	return nil
}

// effectiveLength weighs a character by how much choosing it narrows things
// down.
//
// A Latin letter is one of twenty-six; a hanzi is one of several thousand in
// common use. Counting them the same means "我今天想吃小笼包" — eight
// characters, and a far better password than anything of eight Latin letters —
// gets refused while "aaaaaaaaaab" passes. That is the wrong way round, and it
// lands on exactly the people this system is for.
//
// Two, not the ~2.4 the entropy ratio would justify. The number is a floor
// being set, not a measurement being reported, and rounding it down keeps this
// from becoming a way to pass with three characters.
func effectiveLength(runes []rune) int {
	n := 0
	for _, r := range runes {
		if r > unicode.MaxASCII {
			n += 2
			continue
		}
		n++
	}
	return n
}

// keyboardWalks are the paths a finger takes when somebody is not choosing a
// password so much as declining to.
//
// Checked as substrings rather than listed as literals, because the family is
// generated, not enumerated: 1q2w3e4r, 1q2w3e4r5t, q1w2e3 and a dozen more are
// all the same two rows being zipped together, and a fixed list is a game of
// catch-up that the list always loses.
var keyboardWalks = []string{
	"1234567890", "0987654321",
	"qwertyuiop", "poiuytrewq",
	"asdfghjkl", "lkjhgfdsa",
	"zxcvbnm", "mnbvcxz",
	"1qaz2wsx3edc4rfv5tgb6yhn7ujm",
	"1q2w3e4r5t6y7u8i9o0p",
	"q1w2e3r4t5y6u7i8o9p0",
	"!@#$%^&*()",
}

// walksTheKeyboard reports whether the whole password is a slice of one of
// those paths.
func walksTheKeyboard(password string) bool {
	lower := strings.ToLower(password)
	if len([]rune(lower)) < 6 {
		return false
	}
	for _, walk := range keyboardWalks {
		if strings.Contains(walk, lower) {
			return true
		}
	}
	return false
}

// isCommonPassword strips the decoration people add to a bad password and
// checks what is underneath.
//
// "Password" and "password123!" are the same password wearing different hats,
// and an attacker's list contains both forms. Checking only the literal string
// would let every one of them through.
func isCommonPassword(password string) bool {
	base := strings.ToLower(strings.TrimSpace(password))
	if commonPasswords[base] {
		return true
	}
	// Trailing digits and punctuation: the "add a 1 and a bang" reflex.
	stripped := strings.TrimRight(base, "0123456789!@#$%^&*_-.")
	if stripped != base && commonPasswords[stripped] {
		return true
	}
	// Leading decoration too, less common but the same idea.
	return commonPasswords[strings.Trim(base, "0123456789!@#$%^&*_-.")]
}

// isRepetitive catches "aaaaaaaaaa" and "abcdefghij" — strings long enough to
// pass the length check while containing almost no choice.
//
// Only runs of six or more, so an ordinary password with "789" or "aaa" inside
// it is untouched. The test is on the whole string being one run, not on any
// run existing.
func isRepetitive(runes []rune) bool {
	same, ascending, descending := 1, 1, 1
	for i := 1; i < len(runes); i++ {
		switch {
		case runes[i] == runes[i-1]:
			same++
		default:
			same = 1
		}
		switch {
		case runes[i] == runes[i-1]+1:
			ascending++
		default:
			ascending = 1
		}
		switch {
		case runes[i] == runes[i-1]-1:
			descending++
		default:
			descending = 1
		}
		if same >= 6 || ascending >= 6 || descending >= 6 {
			return true
		}
	}
	return false
}

// borrowsFromIdentity reports which piece of the person's own identity the
// password contains, or "" if none.
//
// Returns the offending word rather than a boolean so the message can name it.
// "密码太弱" sends somebody away guessing; "不要包含「zhangsan」" is a thing
// they can act on in one try.
func borrowsFromIdentity(password string, context []string) string {
	lower := strings.ToLower(password)
	for _, raw := range context {
		for _, part := range identityParts(raw) {
			// Four is the shortest fragment worth refusing. Below that,
			// ordinary words collide with initials and short surnames, and
			// refusing an otherwise good password over "sun" would teach
			// nothing except that this system is annoying.
			if len(part) >= 4 && strings.Contains(lower, part) {
				return part
			}
		}
	}
	return ""
}

// identityParts breaks an address or a name into the fragments somebody would
// actually build a password out of: the part before the @, each label of the
// domain, and each run of letters in a name.
func identityParts(raw string) []string {
	raw = strings.ToLower(strings.TrimSpace(raw))
	if raw == "" {
		return nil
	}
	var out []string
	for _, chunk := range strings.FieldsFunc(raw, func(r rune) bool {
		return !unicode.IsLetter(r) && !unicode.IsDigit(r)
	}) {
		// Public suffixes are shared by everyone and carry no information
		// about this person; refusing a password for containing "com" would
		// be noise.
		switch chunk {
		case "com", "cn", "net", "org", "edu", "gov", "www", "mail", "email":
			continue
		}
		out = append(out, chunk)
	}
	return out
}

File: frontend/src/lib/passwordPolicy.ts
// 密码规则，和服务端那份（services/iam/internal/app/passwordpolicy.go）逐条对应。
//
// 为什么要在浏览器里再写一遍：从前人只能提交了才知道密码不合格，改一次试
// 一次。规则本身不是秘密（它拦的是弱密码，不是拦不知道规则的人），写在这边
// 就能边打边说哪几条过了。
//
// 两份实现会走散，所以有一份**共用的样例**（passwordPolicy.cases.json）：
// 这边的单测和 Go 那边的测试读的是同一个文件，谁改了规则忘了改另一边，
// 两边有一边会红。
//
// **这里没有「必须含大写字母/数字/特殊字符」那种规则**，服务端也没有，是
// 有意的：那套要求逼出来的是 P@ssw0rd1 这种全公司一个模子的密码。把住长度、
// 再把攻击者真正会试的那些挡掉，比凑字符类管用。理由写在 Go 那个文件的
// 开头。

/** 一条没过的规则。顺序和服务端一致：先短，再长，再常见，再规律，最后身份。 */
export type PasswordProblem = 'tooShort' | 'tooLong' | 'tooCommon' | 'tooSimple' | 'fromIdentity'

/** 界面上那份清单里的一条。 */
export interface PasswordRule {
  key: 'length' | 'notCommon' | 'notRepetitive' | 'notIdentity'
  ok: boolean
  /** 没过时的补充，比如密码里撞上的是哪个词。 */
  detail?: string
}

// 十位，其中中文一个字顶两位（见 Go 那边的 effectiveLength：一个汉字是几千
// 选一，一个拉丁字母是二十六选一，按字符数一刀切会专门为难写中文的人）。
const MIN_EFFECTIVE = 10
/** 再怎么"一个字顶两位"，也要有这么多个字符。 */
const MIN_RUNES = 6
/** 上限只为挡住"把一兆文本推进哈希"，不是为了限制长密码。 */
const MAX_LEN = 256
/** 身份片段短于这个就不拦：三个字母的片段和普通单词撞得太多。以字节算，同 Go。 */
const MIN_IDENTITY_PART_BYTES = 4
/** 连着这么多个相同或连续的字符就算太规律。 */
const RUN_LIMIT = 6

const COMMON_PASSWORDS = new Set([
  'password', 'passwd', 'pass', 'admin', 'administrator',
  'root', 'test', 'guest', 'user', 'login',
  'welcome', 'letmein', 'monkey', 'dragon', 'master',
  'qwerty', 'qwertyuiop', 'asdfgh', 'asdfghjkl', 'zxcvbn',
  '1qaz2wsx', '1q2w3e4r', '1q2w3e', 'qazwsx', 'qweasd',
  '123456', '1234567', '12345678', '123456789', '1234567890',
  '111111', '000000', '666666', '888888', '123123',
  'abc123', 'a123456', '123abc', 'abcd1234', 'qwe123',
  'woaini', 'woaini1314', '5201314', '1314520', 'wangyi',
  'iloveyou', 'sunshine', 'princess', 'football', 'baseball',
  'changeme', 'secret', 'default', 'temp', 'demo',
])

const KEYBOARD_WALKS = [
  '1234567890', '0987654321',
  'qwertyuiop', 'poiuytrewq',
  'asdfghjkl', 'lkjhgfdsa',
  'zxcvbnm', 'mnbvcxz',
  '1qaz2wsx3edc4rfv5tgb6yhn7ujm',
  '1q2w3e4r5t6y7u8i9o0p',
  'q1w2e3r4t5y6u7i8o9p0',
  '!@#$%^&*()',
]

/** 公共后缀谁都有，拿它拦密码只是噪音。 */
const PUBLIC_PARTS = new Set(['com', 'cn', 'net', 'org', 'edu', 'gov', 'www', 'mail', 'email'])

/** 装饰性的尾巴：「加个 1 加个叹号」那种。 */
const DECORATION = '0123456789!@#$%^&*_-.'

function chars(s: string): string[] {
  return Array.from(s)
}

/** 非 ASCII 的一个字顶两位，同 Go 的 effectiveLength。 */
function effectiveLength(runes: string[]): number {
  let n = 0
  for (const r of runes) n += (r.codePointAt(0) ?? 0) > 127 ? 2 : 1
  return n
}

function byteLength(s: string): number {
  return new TextEncoder().encode(s).length
}

function trimChars(s: string, set: string, from: 'end' | 'both'): string {
  let start = 0
  let end = s.length
  while (end > start && set.includes(s[end - 1])) end--
  if (from === 'both') {
    while (start < end && set.includes(s[start])) start++
  }
  return s.slice(start, end)
}

export function isTooShort(password: string): boolean {
  const runes = chars(password)
  return runes.length < MIN_RUNES || effectiveLength(runes) < MIN_EFFECTIVE
}

export function isTooLong(password: string): boolean {
  return chars(password).length > MAX_LEN
}

/** 常见密码，连同「脱掉装饰之后是常见密码」的那些。 */
export function isTooCommon(password: string): boolean {
  const base = password.trim().toLowerCase()
  if (COMMON_PASSWORDS.has(base)) return true
  const stripped = trimChars(base, DECORATION, 'end')
  if (stripped !== base && COMMON_PASSWORDS.has(stripped)) return true
  if (COMMON_PASSWORDS.has(trimChars(base, DECORATION, 'both'))) return true
  return walksTheKeyboard(base)
}

function walksTheKeyboard(lower: string): boolean {
  if (chars(lower).length < 6) return false
  return KEYBOARD_WALKS.some((walk) => walk.includes(lower))
}

/** 连着六个相同、或者六个顺着走的（含倒着走）。 */
export function isTooSimple(password: string): boolean {
  const runes = chars(password)
  let same = 1
  let ascending = 1
  let descending = 1
  for (let i = 1; i < runes.length; i++) {
    const cur = runes[i].codePointAt(0) ?? 0
    const prev = runes[i - 1].codePointAt(0) ?? 0
    same = cur === prev ? same + 1 : 1
    ascending = cur === prev + 1 ? ascending + 1 : 1
    descending = cur === prev - 1 ? descending + 1 : 1
    if (same >= RUN_LIMIT || ascending >= RUN_LIMIT || descending >= RUN_LIMIT) return true
  }
  return false
}

/**
 * 密码里撞上的那个身份片段，没有就是空串。
 *
 * context 里允许有 undefined 和空串：调用方往往是直接把「姓名、工号、用户名、
 * 邮箱」几个字段摊进来，而那几个字段常常还没填。让调用方各自过滤，等于把
 * 同一句 filter 抄到每个页面。
 */
export function borrowedFromIdentity(password: string, context: readonly (string | undefined)[]): string {
  const lower = password.toLowerCase()
  for (const raw of context) {
    if (!raw) continue
    for (const part of identityParts(raw)) {
      if (byteLength(part) >= MIN_IDENTITY_PART_BYTES && lower.includes(part)) return part
    }
  }
  return ''
}

function identityParts(raw: string): string[] {
  const clean = raw.trim().toLowerCase()
  if (!clean) return []
  return clean.split(/[^\p{L}\p{N}]+/u).filter((part) => part !== '' && !PUBLIC_PARTS.has(part))
}

/**
 * 第一条没过的规则，顺序和服务端一致。全过回 null。
 *
 * context 是这个人身上攻击者已经知道的那些（姓名、工号、用户名、邮箱）。
 */
export function checkPassword(password: string, context: readonly (string | undefined)[] = []): PasswordProblem | null {
  if (isTooShort(password)) return 'tooShort'
  if (isTooLong(password)) return 'tooLong'
  if (isTooCommon(password)) return 'tooCommon'
  if (isTooSimple(password)) return 'tooSimple'
  if (borrowedFromIdentity(password, context)) return 'fromIdentity'
  return null
}

/**
 * 界面上那份清单：每一条过没过，各算各的。
 *
 * 和 checkPassword 不同——那个到第一条没过就停（要和服务端报同一句话），
 * 这个要把四条都算出来，人才看得见还差哪几条。
 */
export function passwordRules(password: string, context: readonly (string | undefined)[] = []): PasswordRule[] {
  const borrowed = borrowedFromIdentity(password, context)
  return [
    { key: 'length', ok: !isTooShort(password) && !isTooLong(password) },
    { key: 'notCommon', ok: !isTooCommon(password) },
    { key: 'notRepetitive', ok: !isTooSimple(password) },
    { key: 'notIdentity', ok: borrowed === '', detail: borrowed || undefined },
  ]
}

/** 清单上一条的样子：还没开始查（○）、过了（✓）、没过（✕）。 */
export type RuleState = 'idle' | 'ok' | 'bad'

export interface RuleView {
  key: PasswordRule['key']
  state: RuleState
  /** 没过时的补充，比如密码里撞上的是哪个词。 */
  detail?: string
}

/**
 * 界面上那份清单，连同每条该画什么记号。
 *
 * 一个字都没打时四条全是 idle——不打勾也不打叉。检查从打第一个字才开始：
 * 空字符串确实「不是常见密码」，但在人还没写之前就给它打勾，看着像系统在
 * 夸一个不存在的密码（2026-09-18 老板看了第一版说的）。
 */
export function ruleStates(password: string, context: readonly (string | undefined)[] = []): RuleView[] {
  return passwordRules(password, context).map((r) => ({
    key: r.key,
    state: password === '' ? 'idle' : r.ok ? 'ok' : 'bad',
    detail: password === '' ? undefined : r.detail,
  }))
}
